start updating schema doc (#2086)
arviz-devs/arviz: doc/source/schema/schema.md

Currently there are **two beta implementations** of this design:
* {ref}`ArviZ <homepage>` in **Python** which integrates with:
  - [emcee](https://emcee.readthedocs.io/en/stable/)
  - [PyMC](https://www.pymc.io)
  - [Pyro](https://pyro.ai/) and [NumPyro](https://pyro.ai/numpyro/)
  - [PyStan](https://pystan.readthedocs.io/en/latest/index.html),
    [CmdStan](https://mc-stan.org/docs/cmdstan-guide/index.html) and
    [CmdStanPy](https://mc-stan.org/cmdstanpy/)
  - [TensorFlow Probability](https://www.tensorflow.org/probability)

> * [ArviZ.jl](https://github.com/arviz-devs/ArviZ.jl) in **Julia** which integrates with:

Reviewer: ```suggestion
* [InferenceObjects.jl](https://github.com/arviz-devs/InferenceObjects.jl) in **Julia** used in [ArviZ.jl](https://github.com/arviz-devs/ArviZ.jl), which integrates with:
```

  - [CmdStan.jl](https://github.com/StanJulia/CmdStan.jl), [StanSample.jl](https://github.com/StanJulia/StanSample.jl) and [Stan.jl](https://github.com/StanJulia/Stan.jl)

>   - [Turing.jl](https://turing.ml/dev/) and indirectly any package using [MCMCChains.jl](https://github.com/TuringLang/MCMCChains.jl) to store results

Reviewer: [Soss.jl](https://github.com/cscherrer/Soss.jl) can be added here also.

`InferenceData` stores all quantities that are relevant to fulfilling its goals in different groups. Different groups generally distinguish conceptually different quantities in Bayesian inference, however, convenience in {ref}`creation <creating_InferenceData>` and {ref}`usage <working_with_InferenceData>` of `InferenceData` objects also plays a role. In general, each quantity (such as posterior distribution or observed data) will be represented by several multidimensional labeled variables.

### Rules

> Below are a few rules which should be followed:

Reviewer: ```suggestion
Below are a few rules that should be followed:
```

* Each group should have one entry per variable and each variable should be named.

> * Dimension names `chain`, `draw`, `sample` and `post_pred_id` are reserved for

Reviewer 2: ```suggestion
* Dimension names `chain`, `draw`, `sample`, `posterior_pred_id` and `prior_pred_id` are reserved for
```

Reviewer: One might also want a `pred_id` for `predictions` group. What if `pred_id` was the only reserved name and was used in `predictions`, `posterior_predictions`, and `prior_predictions`?

Author: I like `pred_id` much more! Thanks!

  InferenceData use to indicate sample dimensions.

> * Dimensions in InferenceData (including sample dimensions) should be identified by name only.

Reviewer: What does this mean?

Author: That the dimension called `chain` is identified by this and only this. It's position relative to other dimensions is not relevant

* For groups like `observed_data` or `constant_data`, all sample dimensions can be
  omitted. For groups like `prior`, `posterior` or `posterior_predictive` either `sample` has to be
  present or both `chain` and `draw` dimensions need to be present. Any combinations that follow
  this are valid.
* Dimensions must be named and share name with a coordinate specifying the index values, called coordinate values.
* Coordinate values can be repeated and should not necessarily be numerical values.
* Variables must not share names with dimensions.
* Groups, variables or the InferenceData itself can have arbitrary metadata stored.

### Metadata
No metadata is _required_ to be present in order to be compliant with the InferenceData schema.
However, it is recommended to store the following fields when relevant:

> * `name`: InferenceData objects represent multiple quantities related to Bayesian modelling,

Reviewer: Would `model_name` be more explicit that this is meant to be the name of the model, not the name of the group or dataset?

Reviewer: Or is this metadata at the InferenceData level instead of at the group level?

Author: I was thinking about group level attributes here. I'll update the title. Maybe also add a note that the creation_library attrs might be stored in groups (they could be different between groups even though I don't expect this to be common)

  but they are all tied to a single model. The model identifier can be added as metadata
  to simplify the calls to model comparison functions.
* `created_at`: the date of creation of the group.
* `arviz_version`: the version of the ArviZ library that generated the InferenceData

> * `arviz_language`: the programming language from which ArviZ was used to create the InferenceData

Reviewer 2: Could these be a bit more general?

If another program creates the InferenceData objects, it would be a good idea to save its version, language. So also the library name for it might be a good idea.

`creation_library`

Reviewer: I 2nd this, along with a `creation_library_version`. We might also want `creation_library_language`. On the Julia side we could append a `".jl"` to the name if a user provides a package object itself instead of just the name to differentiate, but `".py"` for Python makes less sense.

Author: Yes, that is indeed much better, I'll update it

* `inference_library`: the library used to run the inference.
* `inference_library_version`: version of the inference library used.

### Relations between groups

> `InferenceData` data objects contain any combination of the groups described below. There are also some relations (detailed below) between the variables and dimensions of different groups. Hence, whenever related groups are present they should comply with these relations. Neither the presence of groups nor described below or the lack of some of the groups described below go against the schema.

Reviewer: ```suggestion
`InferenceData` data objects contain any combination of the groups described below. There are also some relations (detailed below) between the variables and dimensions of different groups. Hence, whenever related groups are present they should comply with these relations. Neither the presence of groups not described below or the lack of some of the groups described below go against the schema.
```

#### `posterior`

> Samples from the posterior distribution p(theta|y) in the parameter (also called constrained) space.

Reviewer: Should this use mathjax?

Author: yep, the docs already has mathjax loaded and supports the $ syntax

(schema/unconstrained_posterior)=
#### `unconstrained_posterior`
Samples from the posterior distribution p(theta_transformed|y) in the unconstrained (also called transformed) space.

> Only variables that undergo a transformation for sampling should be present here.

Reviewer 2: This might be a problematic for Stan where it is easy to calculate the transformation but hard to make sure that transformation is not unity (or is comparing values against each other valid way to do this?)

Author: This is only relative to the samples, so comparing values against each other is perfeclty fine. Even in edgecases like a transformation that is the unity in most of the space but not everywhere, if all the samples are the same then the transformation is not doing anything, we don't need to save the samples in both places; and if we generate new posterior samples, we also need to generate the associated unconstrained samples both of which are generated by the PPLs what will know if and which transformations to do.

Maybe we can write it in a bit more clear phrasing. Not the good style for the doc, but in a conversational style it could be something like: I have the posterior samples and I want to have the unconstrained samples. What are the variables whose
samples I need to replace with different ones? -> those variables and samples are the ones that go in this `unconstrained_sample` group.

Therefore, to get the samples for _all_ the variables in the unconstrained space,
variables should be taken from the `unconstrained_posterior` group if present,
and if not, then the values from the variable in the `posterior` group should be used.

> Both samples and variables (for those present only) should match between the `posterior` and the `unconstrained_posterior` groups. Note that as defined above, matching samples and variables impose constraints on dimensions and coordinates, not on the values which will be different.

Reviewer 2: What are these constraints?

E.g. constrained and unconstrained spaces might have different number of dimensions

Author: I was completely unaware of that 😓, so they would have the same sampling dimensions and variable name but not necessarly the same coords and dims, is that right? Would it be ok to allow for variables in unconstrained_posterior to have different shape than that same variable in posterior group, but that if they do have the same shape, then dims and coords should be shared?

Reviewer 2: That sounds like it should work

e.g https://mc-stan.org/docs/reference-manual/simplex-transform.html

cc @avehtari 

Aki do you remember if any constrained vs unconstrained transform had more complicated structure?

Reviewer 3: > Aki do you remember if any constrained vs unconstrained transform had more complicated structure?

Correlation matrices, Covariance matrices, Cholesky factors of covariance matrices, and Cholesky factors of correlation matrices https://mc-stan.org/docs/reference-manual/variable-transforms.html

:::{note}
:class: dropdown

Both InferenceData groups and variables can have metadata, which in the `unconstrained_posterior`
case could be used to store the transformations each variable goes through to map between the
constrained and unconstrained spaces. The schema leaves this completely up to the user

> and imposes on conventions or restrictions on such metadata.

Reviewer 2: ```suggestion
and imposes no conventions or restrictions on such metadata.
```

?

Author: yes

`InferenceData` stores all quantities that are relevant to fulfilling its goals in different groups. Different groups generally distinguish conceptually different quantities in Bayesian inference, however, convenience in {ref}`creation <creating_InferenceData>` and {ref}`usage <working_with_InferenceData>` of `InferenceData` objects also plays a role. In general, each quantity (such as posterior distribution or observed data) will be represented by several multidimensional labeled variables.

### Rules
Below are a few rules that should be followed:
* Each group should have one entry per variable and each variable should be named.

> * Dimension names `chain`, `draw`, `sample` and `pred_id` are reserved for

Reviewer: Oh, you know, `pred_id` is not defined anywhere. IIRC, we discussed using this when one needs multiple IID predictions per sample. Perhaps add a note about this to `posterior_predictive`?

Author: will do, thanks for pointing it out

* Coordinate values can be repeated and should not necessarily be numerical values.
* Variables must not share names with dimensions.
* Groups, variables or the InferenceData itself can have arbitrary metadata stored.

### Metadata
No metadata is _required_ to be present in order to be compliant with the InferenceData schema.
However, it is recommended to store the following fields when relevant:
* `name`: InferenceData objects represent multiple quantities related to Bayesian modelling,
  but they are all tied to a single model. The model identifier can be added as metadata
  to simplify the calls to model comparison functions.
* `created_at`: the date of creation of the group.
* `creation_library`: the library used to create the InferenceData (might not necessarly be ArviZ)
* `creation_library_version`: the version of `creation_library` that generated the InferenceData
* `creation_library_language`: the programming language from which `creation_library` was used to create the InferenceData
* `inference_library`: the library used to run the inference.
* `inference_library_version`: version of the inference library used.

> Metadata can be stored at the whole `InferenceData` level but also at group level when needed.

Reviewer: Do we have any recommendations yet as to which metadata fields belong in groups vs `InferenceData`?

Author: not yet, I'll open an issue for this

the accepted proposal.
* `max_energy_error`: The maximum absolute difference in Hamiltonian energy between the initial point and all possible samples in the proposed tree.
* `int_time`: The total integration time (static HMC sampler)

> * `mass_matrix`: Mass matrix (also known as _metric_) used in HMC samplers for the computation of the Hamiltonian.

Reviewer: Isn't it more common to represent the inverse of the metric instead of the metric itself? I believe this is what Stan does, and this is also what Julia's HMC samplers do. In many/most cases the inverse metric is diagonal, so perhaps we should explicitly say that diagonal metrics can be represented as just the diagonal?

Personally I prefer the term metric over mass matrix.

Reviewer 2: Stan also prefers metric (mass_matrix comes from physics)

Author: I'll update this then, it looks like using `inv_metric` would be best right? also feel free to suggest better definitions or editing it directly

:class: note, dropdown

The schema does not define which warmup or unconstrained groups exist or can exist
by default. We recognize both the samplers and the models are continuously evolving.
Some models already require the use of sampling algorithms to get prior samples,
in which case we basically need to treat the prior and posterior groups in the same way.

We define the prefixes and to allow libraries that use InferenceData to be aware
of the potential relations and hopefully support as many cases as possible.
Back to the case above, it might be necessary to generate a pair plot
for prior samples generated with NUTS _and_ its associated divergences,
which would then come from `sample_stats_prior`.
:::

#### Sample stats groups
Information and diagnostics for the samples in any InferenceData group
other than the posterior should be stored in a separate group with the
`sample_stats_` prefix. For example `sample_stats_prior`.

The same rules and conventions defined in {ref}`schema/sample_stats` apply to
any sample stats group.

#### Warmup groups
Samples generated during the adaptation/warmup phases of algorithms like HMC

> can also be stored in InferenData. In such cases, the data/samples

Reviewer 2: ```suggestion
can also be stored in InferenceData. In such cases, the data/samples
```

generated during the adaptation process should be stored in groups with

> the same name with the `warmup_` prefix, e.g. `warmup_posterior`, `warmup_sample_stats_prior`.

Reviewer 2: How would the naming go for unconstained warmup groups

    warmup_unconstrained_posterior

?

Author: yes, I would make the `warmup_` prefix prepend the other two potential prefixes. I'll add that.

#### `predictions_constant_data`
Model constants used to get the `predictions` samples. Its variables should have a counterpart in `constant_data`. However, variables in `predictions_constant_data` and their counterpart in `constant_data` can have different coordinate values.

:::{admonition} Note on sample stats, warmup and unconstrained groups
:class: note, dropdown

The schema does not define which warmup or unconstrained groups exist or can exist
by default. We recognize both the samplers and the models are continuously evolving.
Some models already require the use of sampling algorithms to get prior samples,
in which case we basically need to treat the prior and posterior groups in the same way.

> We define the prefixes and to allow libraries that use InferenceData to be aware

Reviewer: ```suggestion
We define the prefixes to allow libraries that use InferenceData to be aware
```
